### ISSUE TITLE
Huge memory leak makes whole computer to stop

### ISSUE THREAD

**Issue by glaubersm** [flathub/com.moddb.TotalChaos#6]
Hi
I trying to run total chaos on Arch Linux running X11 session of KDE Plasma 5.19.4 desktop environment
but it is unusable because an huge memory leak that begins when I start a new game and makes the whole computer to stop.

![memoryleak](https://user-images.githubusercontent.com/15967631/91579941-bac73a00-e922-11ea-81bc-744375146dca.png)

**Comment by AsciiWolf** [comment]
Hi, hmm, that's weird. Yeah, the game is memory intensive, but you are right that such memory allocation should not happen. Unfortunately, I have no machine with only 8GB of RAM so I can't test it on such lower-end system. Let's hope that the new [Director's Cut edition](https://github.com/flathub/com.moddb.TotalChaos/issues/5) (that I plan to update to) with compressed textures and other memory optimisations will at least partially resolve these issues...

**Comment by AsciiWolf** [comment]
Hello @glaubersm, a new Total Chaos Flatpak version (1.32) was just released. Please, let me know if you are still able to reproduce this issue. Thanks!

**Comment by glaubersm** [comment]
The problem persists.

**Comment by AsciiWolf** [comment]
Total Chaos Flatpak just got updated along with GZDoom and other dependencies. The update should be live on Flathub in a few hours. Please, retest and let me know whether the problem still persists. Thanks!

**Comment by glaubersm** [comment]
Currently my pc has 16 GiB of RAM and the memory leak persists. Fortunately the current kernel kills the process
and my pc no longer becomes completely unusable.

**Comment by AsciiWolf** [comment]
That's really strange. :-/ I have tried this Flatpak on multiple computers with multiple software and hardware configurations and did not have any issue at all.

**Comment by glaubersm** [comment]
Do your computers have intel iGPU?

**Comment by AsciiWolf** [comment]
One of the machines I have tried Total Chaos on was a ThinkPad laptop with integrated Intel GPU. The game was really slow on that machine, but there was no memory leak.

**Comment by glaubersm** [comment]
Well, what a bizarre problem then. :(
In addition to the memory leak, the game is unusable on my pc due the extremely low framerate.
I do no kwow what are the hardware requeriments.

**Comment by AsciiWolf** [comment]
The game is pretty demanding, especially regarding the GPU, so low framerates on an iGPU setup are pretty normal. However, there should not be any memory leaks.

**Comment by mbugni** [comment]
Hi @glaubersm,
I just released a build to get a [retro version](https://github.com/mbugni/TotalChaosRetro) of the game (just for fun). It has lower resolution, but maybe it can help you. You can install it along with the Flathub's version.

**Comment by AsciiWolf** [comment]
@mbugni Wow, that's cool! Do you plan to release it on Flathub? :-)

**Comment by glaubersm** [comment]
```
Download (inside project directory) the original totalchaos_dc_retro_b5.zip file from:
```
@mbugni 
What is the "project directory"?

**Comment by mbugni** [comment]
> @mbugni Wow, that's cool! Do you plan to release it on Flathub? :-)

Hi @AsciiWolf,  I'm not a game expert or a package maintainer. This is my very first manifest.
I followed your work on regular version (thanks guys!) and added this patch to the retro version:

```yaml
build-commands:
      # Patch shaders to avoid runtime errors
      - unzip totalchaos.pk3 SHADERS/old.fp
      - sed -i 's/texture2D/texture/g' SHADERS/old.fp
      - zip -u totalchaos.pk3 SHADERS/old.fp
```

But I'm not sure to understand the reason, just googled about it.
If you think is feasible, I'll be glad to collaborate.

**Comment by mbugni** [comment]
> ```
> Download (inside project directory) the original totalchaos_dc_retro_b5.zip file from:
> ```
> 
> @mbugni What is the "project directory"?

It's the directory where you dowloand the project files, the one which contains the `io.github.mbugni.TotalChaosRetro.yaml` file.

**Comment by mbugni** [comment]
FYI 

I tried to build the retro edition on my 10+ years old laptop, but it didn't work because GZDoom 4 requires OpenGL 3 or later and my laptop supports only version 2.1.
So I decided to use the GZDoom v3.8.2 legacy version, which supports older OpenGL.
Now the game run fast(!) and smoothly(!) even on my old PC, which has:
- CPU: Intel Core i3 350M 2.27GHz
- GPU: Intel Integrated Graphics
- RAM: 4GB DDR3 
- Screen resolution: 1366x768 pixels
- OpenGL: 2.1 Mesa 22.3.7
- KDE X11 on Fedora 37

I think GZDoom legacy engine is definitely the best choice for retro edition, so I decided to modify my build to use it.
Hope it helps.

**Comment by glaubersm** [comment]
I have installed Total Chaos Retro from flathub and I'm unable to reproduce the memory leak with it.
However, mouse input is not working.

**Comment by mbugni** [comment]
Just this app? All flatpak apps? Flatpak version?

@glaubersm if the mouse problem persist, I think we can move the discussion on TotalChaosRetro repo.
@AsciiWolf I think the memory problem has gone with the lighter version.

**Comment by glaubersm** [comment]
Mouse works with other games from flathub:
Total Chaos
Nexuiz Classic
Unvanquished

I use Arch Linux running KDE Plasma 5.27.5 and flatpak 1:1.15.4-1.

**Comment by mbugni** [comment]
Stupid question, anyway: is mouse enabled (Options > Mouse Options > Enable mouse > Yes) ?

**Comment by AsciiWolf** [comment]
I have just tried the Flathub build of Total Chaos Retro and mouse works fine in game on my system.

**Comment by glaubersm** [comment]
"Enable mouse in menus" was disabled. I have enabled it and now the mouse works. :)
Thanks.

**Comment by PavanLuca** [comment]
I'm not sure if it's related but in my case it doesn't stop the whole computer, it just hangs some minutes during map loading and then crashes. On my logs software I see a memory outage, but I solve the problem by setting one of the other 2 renderers instead of hardware acceleration. But then despite the game is finally playable the problem is that it loads some yellow and green squares on certain points of textures.

**Comment by PavanLuca** [comment]
I'm not sure if it's related but in my case it doesn't stop the whole computer, it just hangs some minutes during map loading and then crashes. On my logs software I see a memory outage, but I solve the problem by setting one of the other 2 renderers instead of hardware acceleration. But then despite the game is finally playable the problem is that it loads some yellow and green squares on certain points of textures. And moreover certain settings aren't saved once restarting the game.

**Comment by glaubersm** [comment]
Can confirm. Currently the computer remains usable and the game crashes.

**Comment by PavanLuca** [comment]
To those whose game crashes try one of the other 2 renderers as I did. Also, if this way it works please tell me if you see paranormal squares on some textures.

**Comment by glaubersm** [comment]
doom and true color software rendering work, can confirm squares with true color render.

**Comment by PavanLuca** [comment]
Thanks for telling it, but I have squares on both these 2 renderers.

**Comment by PavanLuca** [comment]
Thanks for telling it, but I have squares on both these 2 renderers. To whom tried the Retro Edition, did you notice this bug either, some others or none?

**Comment by AsciiWolf** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1691

Related?

**Comment by PavanLuca** [comment]
Yes, seems that, it's mentioned AMD like mine in the flat version of course. Anyway I uninstalled the game after trying different configurations concerning the rendering and other “heavy settings” and getting no particular help from the official Discord server. Changing the renderer made the game playable but at the cost of showing cubes in certain spots and not loading textures to certain objects (like doors and keys, I found a key on the room upstairs on the first map by luck, I was just clicking around 😄) I decided to stop it. Moreover there was a known bug that was making certain configurations to be discarded once the game was closed and this problem plus the previous 2 ones made me lose the interest.

**Comment by souxd** [comment]
some years ago, before the flatpak existed, i tried this game and had the same problem. I've tried it both on an intel hd 2000 and a polaris card. Both had the memleak and i couldn't play because of it. Today i've tried again on the polaris card through this flatpak and it only works on software mode like Pavan has said. I suspect the issue is in GZDoom but i couldn't find anything related on their github

**Comment by mbugni** [comment]
The engine is [moving to UZDoom](https://github.com/flathub/com.moddb.TotalChaos/pull/27).

**Comment by mbugni** [comment]
@glaubersm @souxd @PavanLuca is this still an issue after the UZDoom engine adoption?

**Comment by souxd** [comment]
I had already tested the flatpak on the new commit and it worked fine. I was just waiting for OP.

**Comment by mbugni** [comment]
Closing this, since the new engine is online and it seems to work better.
